code/game/objects/machinery/miner.dm

```diff
@@ -43,21 +50,84 @@
-			icon_state = "mining_drill"
+			icon_state = "mining_drill_[miner_upgrade_type]"
 		if(MINER_DESTROYED)
-			icon_state = "mining_drill_error"
+			icon_state = "mining_drill_error_[miner_upgrade_type]"
+/// Called whenever someone attacks the miner with a object which is considered a upgrade.The object needs to have a uptype var.
+/obj/machinery/miner/proc/attempt_upgrade(obj/item/minerupgrade/upgrade, mob/user, params)
+	if(miner_upgrade_type)
+		to_chat(user, "<span class='info'>The [src]'s module sockets are already occupied by the [miner_upgrade_type].</span>")
+		return FALSE
+	if(user.skills.getRating("construction")<SKILL_CONSTRUCTION_ADVANCED)
+		to_chat(user, "<span class='info'>You can't figure out how to install the complex module.</span>")
+		return FALSE
+	user.visible_message("<span class='notice'>[user] begins attaching a module to [src]'s sockets.</span>")
+	to_chat(user, "<span class='info'>You begin installing the [upgrade] on the miner.</span>")
+	if(!do_after(user, 15 SECONDS, TRUE, src, BUSY_ICON_BUILD))
+		return FALSE
+	switch(upgrade.uptype)
+		if(MINER_RESISTANT)
+			max_miner_integrity = 300
+			miner_integrity = 300
+		if(MINER_COMPACTOR)
+			if(mineral_produced == /obj/structure/ore_box/platinum)
+				mineral_produced = /obj/item/compactorebox/platinum
+			else
+				mineral_produced = /obj/item/compactorebox/phoron
+		if(MINER_OVERCLOCKED)
+			required_ticks = 35
+	miner_upgrade_type = upgrade.uptype
+	user.visible_message("<span class='notice'>[user] attaches the [miner_upgrade_type] to the [src]!</span>")
+	qdel(upgrade)
+	playsound(loc,'sound/items/screwdriver.ogg', 25, TRUE)
+	update_icon()
+
+/obj/machinery/miner/attackby(obj/item/I,mob/user,params)
+	. = ..()
 
+	if(istype(I, /obj/item/minerupgrade))
+		var/obj/item/minerupgrade/upgrade = I
+		if(!(miner_status == MINER_RUNNING))
+			to_chat(user, "<span class='info'>[src]'s module sockets seem bolted down.</span>")
+			return FALSE
+		attempt_upgrade(upgrade,user)
 
 /obj/machinery/miner/welder_act(mob/living/user, obj/item/I)
 	. = ..()
+	var/obj/item/tool/weldingtool/weldingtool = I
+	if((miner_status == MINER_RUNNING) && miner_upgrade_type)
+		if(!weldingtool.remove_fuel(2,user))
+			to_chat(user, "<span class='info'>You need more welding fuel to complete this task!</span>")
+			return FALSE
+		to_chat(user, "<span class='info'>You begin uninstalling the [miner_upgrade_type] from the miner</span>")
+		user.visible_message("<span class='notice'>[user] begins dismantling the [miner_upgrade_type] from the miner</span>")
+		if(!do_after(user, 30 SECONDS, TRUE, src, BUSY_ICON_BUILD))
+			return FALSE
+		user.visible_message("<span class='notice'>[user] dismantles the [miner_upgrade_type] from the miner</span>")
+		var/obj/item/upgrade
+		switch(miner_upgrade_type)
+			if(MINER_RESISTANT)
+				upgrade = new /obj/item/minerupgrade/reinforcement
+				max_miner_integrity = initial(max_miner_integrity)
+				miner_integrity = initial(miner_integrity)
```

Comment:
This needs a check for whatever the current HP is. if less than the initial, no change to current HP.

Reply:
The proc used to upgrade doesn't allow upgrading a damaged drill.

Reply:
added in a check